### DISCUSSION TITLE
Automatic Login via shell ?

### DISCUSSION THREAD

**Discussion by yaq-cc** [coder/code-server#5274] [General]
Hi code-server team,

This is about some linux oriented scripting automation.  My code-server instance runs on a pre-emptible compute engine instance.  Because of this I want to provide a shutdown script which uses linux's wall program to alert all logged in users that a pre-emption request was issued.  

wall "Pre-emption detected.  Please save your work!" 

For the wall command to work, the user must set mesg y and must be logged in to the shell.  The setting of mesg to y is done via the user's ~/.profile.  My current problem is that I can't script the "login".  I tried adding a sudo login -f $USER but this didn't work.  instead it logged the user into the same shell over and over and over nonstop.

I have two questions:

(a) How would I go about logging a shell user in automatically and 
(b) How can I perform a shell "teardown" such that if I close a shell via exit or closing the shell the user is then logged out of that shell..?

Thank you!

**Comment by jsjoeio** [comment]
Hmm... I don't know if you can script the login 🤔 Only thought that comes to mind is disabling the login with `--auth none` but that's probably not what you want.

cc @oxy @code-asher any ideas?

**Comment by code-asher** [comment]
I have not tested any of this but here are my thoughts.

> (a) How would I go about logging a shell user in automatically 

You could use your current method but detect if you are already in a login shell in order to avoid the loop.

Bash will set `login_shell` but other shells may be different:

```
if ! shopt -q login_shell ; then
  sudo login -f "$USER"
fi
```

Another idea is to change a setting in code-server that makes the shell spawn as a login shell by adding a flag.  For Bash this flag is `--login` (or `-l`).  For example by adding this to the settings file:

```
"terminal.integrated.shellArgs.linux": ["--login"]
```

>  (b) How can I perform a shell "teardown" such that if I close a shell via exit or closing the shell the user is then logged out of that shell..?

If I understand correctly you want `exit` to close out completely rather than dropping back to the non-login shell?  

If you use the first method above you can add `exec` which will replace the current shell (`exec sudo login -f "$USER"`).

If you use the second method nothing further needs to be done.



>

**Reply by code-asher** [reply]
I had the `if` reversed, edited above but here it is for any email users:

```
if ! shopt -q login_shell ; then
  sudo login -f "$USER"
fi
```
